File: VonNeumannModel.cpp
#include <iostream>
#include <vector>

// Memory
class Memory {
    std::vector<int> data;

public:
    Memory(size_t size) : data(size, 0) {}

    int operator[](size_t address) {
        return data[address];
    }

    void set(size_t address, int value) {
        data[address] = value;
    }
};

// CPU
class CPU {
    Memory& memory;
    int acc;  // accumulator
    size_t pc;  // program counter

public:
    CPU(Memory& memory) : memory(memory), acc(0), pc(0) {}

    int fetch() {
        int opcode = memory[pc];
        pc += 1;
        return opcode;
    }

    void execute(int opcode) {
        if (opcode == 0) {  // LOAD
            size_t address = memory[pc];
            pc += 1;
            acc = memory[address];
        } else if (opcode == 1) {  // STORE
            size_t address = memory[pc];
            pc += 1;
            memory.set(address, acc);
        }
        // implement other opcodes...
    }

    void run() {
        while (true) {
            int opcode = fetch();
            execute(opcode);
        }
    }
};

int main() {
    Memory memory(256);
    CPU cpu(memory);

    // A simple program: load the number 42 into the accumulator, then store it at memory location 0
    memory.set(0, 0);  // LOAD
    memory.set(1, 1);  // memory location 1
    memory.set(2, 1);  // STORE
    memory.set(3, 0);  // memory location 0
    memory.set(1, 42);  // number to load

    cpu.run();

    return 0;
}

File: VonNeumannModel.py
class Memory:
    def __init__(self, size):
        self.data = [0] * size

    def __getitem__(self, address):
        return self.data[address]

    def __setitem__(self, address, value):
        self.data[address] = value

class CPU:
    def __init__(self, memory):
        self.memory = memory
        self.acc = 0  # accumulator
        self.pc = 0  # program counter

    def fetch(self):
        opcode = self.memory[self.pc]
        self.pc += 1
        return opcode

    def execute(self, opcode):
        if opcode == 0:  # LOAD
            address = self.memory[self.pc]
            self.pc += 1
            self.acc = self.memory[address]
        elif opcode == 1:  # STORE
            address = self.memory[self.pc]
            self.pc += 1
            self.memory[address] = self.acc
        # implement other opcodes...

    def run(self):
        while True:
            opcode = self.fetch()
            self.execute(opcode)

memory = Memory(256)
cpu = CPU(memory)

# A simple program: load the number 42 into the accumulator, then store it at memory location 0
memory[0] = 0  # LOAD
memory[1] = 1  # memory location 1
memory[2] = 1  # STORE
memory[3] = 0  # memory location 0
memory[1] = 42  # number to load

cpu.run()
